src/main/java/spliterators/part1/exercise/RectangleSpliterator.java

```diff
@@ -86,11 +94,12 @@ public void forEachRemaining(IntConsumer action) {
             int currentInnerFromInclusive = i == outerFromInclusive ? innerFromInclusive : 0;
             int currentInnerToExlusive = i == outerToExclusive - 1 ? innerToExclusive : innerLength;
 
-            for (int j = currentInnerFromInclusive; j < currentInnerToExlusive; j++) {
+            int j;
+            for (j = currentInnerFromInclusive; j < currentInnerToExlusive; j++) {
                 action.accept(array[i][j]);
-                innerFromInclusive = j;
             }
+            innerFromInclusive = j;

```

Comment:
Use `innerFromInclusive = currentInnerToExlusive` and move `j` declaration into `for`.